Bluetooth: csis: Expose bt_csis_generate_rsi function

MariuszSkamra opened the pull request:
The function will be used to update the RSI value. The returned value
can be later used in advertisement data for set identification. This
removes the need of specific bt_csis_advertise function so that the user
can call the API to generate new RSI once needed (e.g. on RPA change).
This allows the user to manage the advertisement data to fit it's needs.

< HCI Command: LE Set Extended Advertising Enable (0x08|0x0039) plen 6
        Extended advertising: Disabled (0x00)
        Number of sets: 1 (0x01)
        Entry 0
          Handle: 0x00
          Duration: 0 ms (0x00)
          Max ext adv events: 0
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Extended Advertising Enable (0x08|0x0039) ncmd 1
        Status: Success (0x00)
< HCI Command: LE Set Extended Advertising Data (0x08|0x0037) plen 33
        Handle: 0x00
        Operation: Complete extended advertising data (0x03)
        Fragment preference: Minimize fragmentation (0x01)
        Data length: 0x1d
        Flags: 0x06
          LE General Discoverable Mode
          BR/EDR Not Supported
        Unknown EIR field 0x2e: 16e61d64dc45
        Name (complete): audio test shell
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Extended Advertising Data (0x08|0x0037) ncmd 1
	Status: Success (0x00)
< HCI Command: LE Set Advertising Set Random Address (0x08|0x0035) plen 7
        Advertising handle: 0x00
        Advertising random address: 4E:21:29:F8:94:93 (Resolvable)
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Advertising Set Random Address (0x08|0x0035) ncmd 1
        Status: Success (0x00)
< HCI Command: LE Set Extended Advertising Enable (0x08|0x0039) plen 6
        Extended advertising: Enabled (0x01)
        Number of sets: 1 (0x01)
        Entry 0
          Handle: 0x00
          Duration: 0 ms (0x00)
          Max ext adv events: 0
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Extended Advertising Enable (0x08|0x0039) ncmd 1
        Status: Success (0x00)

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: csis: Expose bt_csis_generate_rsi function

The function will be used to update the RSI value. The returned value
can be later used in advertisement data for set identification. This
removes the need of specific bt_csis_advertise function so that the user
can call the API to generate new RSI once needed (e.g. on RPA change).
This allows the user to manage the advertisement data to fit it's needs.

< HCI Command: LE Set Extended Advertising Enable (0x08|0x0039) plen 6
        Extended advertising: Disabled (0x00)
        Number of sets: 1 (0x01)
        Entry 0
          Handle: 0x00
          Duration: 0 ms (0x00)
          Max ext adv events: 0
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Extended Advertising Enable (0x08|0x0039) ncmd 1
        Status: Success (0x00)
< HCI Command: LE Set Extended Advertising Data (0x08|0x0037) plen 33
        Handle: 0x00
        Operation: Complete extended advertising data (0x03)
        Fragment preference: Minimize fragmentation (0x01)
        Data length: 0x1d
        Flags: 0x06
          LE General Discoverable Mode
          BR/EDR Not Supported
        Unknown EIR field 0x2e: 16e61d64dc45
        Name (complete): audio test shell
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Extended Advertising Data (0x08|0x0037) ncmd 1
	Status: Success (0x00)
< HCI Command: LE Set Advertising Set Random Address (0x08|0x0035) plen 7
        Advertising handle: 0x00
        Advertising random address: 4E:21:29:F8:94:93 (Resolvable)
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Advertising Set Random Address (0x08|0x0035) ncmd 1
        Status: Success (0x00)
< HCI Command: LE Set Extended Advertising Enable (0x08|0x0039) plen 6
        Extended advertising: Enabled (0x01)
        Number of sets: 1 (0x01)
        Entry 0
          Handle: 0x00
          Duration: 0 ms (0x00)
          Max ext adv events: 0
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Extended Advertising Enable (0x08|0x0039) ncmd 1
        Status: Success (0x00)

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on include/zephyr/bluetooth/audio/csis.h:
```diff
@@ -190,16 +191,16 @@ int bt_csis_register(const struct bt_csis_register_param *param,
  *
  * @param csis          Pointer to the Coordinated Set Identification Service.
- * @param enable	If true start advertising, if false stop advertising
+ * @param rsi		Pointer to the 6-octet newly generated RSI data.
```
```suggestion
 * @param rsi           Pointer to the 6-octet newly generated RSI data.
```
spaces instead of tabs

Thalley commented on subsys/bluetooth/shell/csis.c:
```diff
@@ -105,20 +105,114 @@ static int cmd_csis_register(const struct shell *sh, size_t argc, char **argv)
 }
 
+static uint8_t rsi[BT_CSIS_RSI_SIZE];
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_CSIS_DATA_RSI(rsi),
+};
```
The CSIP spec states that
> A privacy-enabled Set Member should advertise RSI values only when in the GAP Limited Discoverable mode (see Volume 3, Part C, Section 9.2.3 in [2]).

So perhaps we should reflect that in this adv data as well? i.e. use `BT_LE_AD_LIMITED` instead of `BT_LE_AD_GENERAL` if privacy is enabled.

Thalley reviewed (changes requested):
Minor change request and some comments/questions, otherwise LGTM. 

I think this is the right way to implement this, as the CSIP spec also states 
> If privacy is not enabled, the RSI may not change, or the RSI renewal period may be defined by a higher-layer specification.

So we need to be able allow upper layers to renew the RSI at any time.

MariuszSkamra commented on subsys/bluetooth/shell/csis.c:
```diff
@@ -105,20 +105,114 @@ static int cmd_csis_register(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+
+	err = csis_adv_update_rsi(csis, adv);
+	if (err != 0) {
+		return false;
+	}
```
@Thalley Would it be a good idea to instead of having BT_CSIS_ENC_SIRK_SUPPORT option, rely on BT_PRIVACY instead? So if the privacy is enabled, the SIRK will be encrypted, if disabled, it will be plain text.

MariuszSkamra replied:
> Given that a higher layer profile may require the SIRK to be encrypted, but without privacy, I vote that we enable `BT_CSIS_ENC_SIRK_SUPPORT=y` by default if `BT_PRIVACY=y`, but do not _require_ it (spec uses "should" and not "shall"), but do not otherwise restrict it.

In that case, no changes are needed in `Kconfig.csis`, as `BT_CSIS_ENC_SIRK_SUPPORT` is enabled by default anyway

Thalley commented:
@MariuszSkamra One thing I considered was whether or not we should define the API in a more generic way? 

CSIS has some `service data` that needs to be advertised (the RSI). 

Similarly the BAP Unicast Server has some `service data`, and ditto for the broadcast sink (even though the data isn't really part of a service...), and this is the case for many of our (peripheral) roles. 

Should we consider add a more generic API for these, such that it is
`bt_csip_csis_get_service_data`, `bt_bap_broadcast_sink_get_service_data`, bt_cap_acceptor_get_service_data`, instead of the more specific `bt_csis_generate_rsi`?

MariuszSkamra commented:
> @MariuszSkamra One thing I considered was whether or not we should define the API in a more generic way?
> 
> CSIS has some `service data` that needs to be advertised (the RSI).
> 
> Similarly the BAP Unicast Server has some `service data`, and ditto for the broadcast sink (even though the data isn't really part of a service...), and this is the case for many of our (peripheral) roles.
> 
> Should we consider add a more generic API for these, such that it is `bt_csip_csis_get_service_data`, `bt_bap_broadcast_sink_get_service_data`, bt_cap_acceptor_get_service_data`, instead of the more specific `bt_csis_generate_rsi`?

I agree, that it would be good to have a common approach. Maybe it might be some kind of struct. I would like to merge it as it is and maybe change it later as a part of refactor

MariuszSkamra committed:
Bluetooth: shell: Add LE Audio related advertisement data

This adds LE Audio related data to the shell advertisement data.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
@szymon-czapracki FYI

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1152,85 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+		len++;
+	}
+
+#if defined(CONFIG_BT_AUDIO)
```
Any reason for the `CONFIG_BT_AUDIO` guard here?

MariuszSkamra replied:
it's because e.g. `audio.h` cannot be included because some of the struct members are dependent on Kconfig option that is available only if `CONFIG_BT_AUDIO` is enabled. E.g `bt_codec->data[CONFIG_BT_CODEC_MAX_DATA_COUNT]` Thus I had to guard the includes and some parts of the code.

Thalley replied:
> it's because e.g. audio.h cannot be included because some of the struct members are dependent on Kconfig option that is available only if CONFIG_BT_AUDIO is enabled.

That should be fixed then :) Public header files should always be possible to include regardless of any Kconfig option.

 It's fine to postpone that fix to a later PR and remove this guard, but please add a comment/todo here then.

MariuszSkamra replied:
I'll just add the fix as separate patch in this patchset

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1152,85 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+		/* A privacy-enabled Set Member should only advertise RSI values derived
+		 * from a SIRK that is exposed in encrypted form.
+		 */
+		if (IS_ENABLED(CONFIG_BT_PRIVACY) &&
+		    !IS_ENABLED(CONFIG_BT_CSIS_ENC_SIRK_SUPPORT)) {
+			shell_warn(ctx_shell, "unencrypted SIRK, excluding AD_RSI");
```
Hmm, not really sure about this, as we are basically overwriting a "shall" based on a "should", i.e. 

> When using this profile [CSIP] over the LE transport, Set Members in the GAP Peripheral role shall include the RSI AD Type (defined in [1]) in the Advertising Data or Scan Response Data.

is being blocked by

> A privacy-enabled Set Member should only advertise RSI values derivedfrom a SIRK that is exposed in encrypted form.

I think it is OK to add the warning, but IMO we should still advertise the RSI.

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1152,85 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+			__ASSERT(data_size > len, "No space for AD_RSI");
+			data[len].type = BT_DATA_CSIS_RSI;
+			data[len].data_len = ARRAY_SIZE(ad_rsi);
+			data[len].data = &ad_rsi[0];
```
```suggestion
			data[len].data = ad_rsi;
```

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1152,85 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+#if defined(CONFIG_BT_AUDIO)
+	/* Advertise RSi in discoverable mode only */
+	if (IS_ENABLED(CONFIG_BT_CSIS) && csis != NULL && discoverable) {
+		static uint8_t ad_rsi[BT_CSIS_RSI_SIZE];
```
The scope of this can be reduced

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1152,85 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+
+#if defined(CONFIG_BT_AUDIO)
+	/* Advertise RSi in discoverable mode only */
+	if (IS_ENABLED(CONFIG_BT_CSIS) && csis != NULL && discoverable) {
```
I think this is a decent way of adding service data to the BT shell advertise command, as it can be fairly easily extended. 

The part that is guarded by `IS_ENABLED(CONFIG_BT_CSIS)` could perhaps be moved to a function in the CSIP/CSIS shell module instead and simply called from here

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1424,9 +1501,74 @@ static int cmd_adv_param(const struct shell *sh, size_t argc, char *argv[])
 }
 
+static const uint8_t ad_ext_uuid16[] = {
+	IF_ENABLED(CONFIG_BT_ASCS, (BT_UUID_16_ENCODE(BT_UUID_ASCS_VAL),))
+	IF_ENABLED(CONFIG_BT_BASS, (BT_UUID_16_ENCODE(BT_UUID_BASS_VAL),))
+	IF_ENABLED(CONFIG_BT_PACS, (BT_UUID_16_ENCODE(BT_UUID_PACS_VAL),))
+	IF_ENABLED(CONFIG_BT_CAP_ACCEPTOR, (BT_UUID_16_ENCODE(BT_UUID_CAS_VAL),))
+	IF_ENABLED(CONFIG_BT_HAS, (BT_UUID_16_ENCODE(BT_UUID_HAS_VAL),))
+	IF_ENABLED(CONFIG_BT_VCS, (BT_UUID_16_ENCODE(BT_UUID_VCS_VAL),))
```
I'm curious as to why this specific subset of services were chosen to have their UUID advertised. Can you elaborate?

MariuszSkamra replied:
As written in specifications e.g:
BAP:
![image](https://user-images.githubusercontent.com/10126877/181756452-e77986e3-687f-49f4-a20b-b47046138ef4.png)
HAP:
![image](https://user-images.githubusercontent.com/10126877/181756614-45b7c1a8-2d6e-43f5-b127-90718c175ae2.png)
VCP:
![image](https://user-images.githubusercontent.com/10126877/181758498-0944106e-09ff-4224-9f3f-c9f08d5f6163.png)
MICP:
![image](https://user-images.githubusercontent.com/10126877/181758983-d56e7472-926b-4438-ae91-ffc745115837.png)

Thalley replied:
I get that, but I was wondering why this specific **subset** of the services were chosen, because e.g. CCP has similar text but the TBS service UUID isn't added here. I assume that MICP also includes similar language.

MariuszSkamra replied:
I will extend the data with missing services. I just forgot about those, that's why

Thalley replied:
Alright :D

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1424,9 +1501,74 @@ static int cmd_adv_param(const struct shell *sh, size_t argc, char *argv[])
+		__ASSERT(data_size > len, "No space for AD_UUID16");
+		data[len].type = BT_DATA_UUID16_SOME;
+		data[len].data_len = ARRAY_SIZE(ad_ext_uuid16);
+		data[len].data = &ad_ext_uuid16[0];
```
```suggestion
		data[len].data = ad_ext_uuid16;
```

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1424,9 +1501,74 @@ static int cmd_adv_param(const struct shell *sh, size_t argc, char *argv[])
 	return 0;
 }
 
+static const uint8_t ad_ext_uuid16[] = {
```
Can be moved to `ad_ext_init`

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1424,9 +1501,74 @@ static int cmd_adv_param(const struct shell *sh, size_t argc, char *argv[])
+		__ASSERT(data_size > len, "No space for AD_CAP_ANNOUNCEMENT");
+		data[len].type = BT_DATA_SVC_DATA16;
+		data[len].data_len = ARRAY_SIZE(ad_cap_announcement);
+		data[len].data = &ad_cap_announcement[0];
```
Since this seems to be a style you've chosen, can you explain why you are doing `&ad_cap_announcement[0]` instead of `ad_cap_announcement`? Both ways work of course, but I'm just curious as to why you prefer the former?

MariuszSkamra replied:
I like it better ^^

Thalley replied:
I guess that's a personal observation then :D I have it the other way around, i.e. `data[len].data = ad_cap_announcement` makes me assume that `ad_cap_announcement` is an array (or in any case the data that is being assigned), while `data[len].data = &ad_cap_announcement[0];` indicates, to me, that it's a specific element of an array that is being used. 

But I'll resolve my comments as we don't have a clear guide on this (and I am unaware of any MISRA rules related to this).

jori-nordic commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1160,83 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+		ad_len++;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CSIS)) {
```
this part seems like it should rather be in csis.c or audio.c, same for the externs in bt.h

Thalley replied:
We could move the CSIS stuff into `audio_ad_data_add`. 

CSIS can be considered outside the scope of audio (As it can used for non-audio stuff), so maybe it is better to put into the csis.c shell module instead.

jori-nordic replied:
Yeah I have a bunch to catch up on in the spec. I hope I'll have some time to do that once we've (aleks and I) addressed the more critical host bugs.

MariuszSkamra replied:
So, as I understand, It can be left as it is, since `csis_ad_data_add`called here, is in `csis.c` already

Thalley replied:
Indeed

sjanc commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -259,11 +269,11 @@ struct bt_codec {
 	size_t   data_count;
 	/** Codec Specific Data */
-	struct bt_codec_data data[CONFIG_BT_CODEC_MAX_DATA_COUNT];
+	struct bt_codec_data data[BT_CODEC_MAX_DATA_COUNT];
```
this might result in undefined behavior ( 0 length array inside struct),  I'd just #ifdef member instead

Thalley commented on subsys/bluetooth/shell/audio.c:
```diff
@@ -1575,3 +1575,78 @@ static int cmd_audio(const struct shell *sh, size_t argc, char **argv)
+	}
+
+	if (ARRAY_SIZE(ad_ext_uuid16) > 0) {
+		__ASSERT(data_size > ad_len, "No space for AD_UUID16");
```
Furthermore, the UUIDs of the services is not required (they are a "should"), but the service data (e.g. `ad_cap_announcement` and `ad_bap_announcement`) are "shall"s. 

We should put the service data before the UUIDs so that the UUIDs in the advertising data is not prioritized higher than the service data.

MariuszSkamra replied:
> `ad_len` is always 0 here; was this meant to be in a loop?

`ad_len` is 0, but `data_size` is provided by user, which might be 0, so the check is still valid

> We should put the service data before the UUIDs so that the UUIDs in the advertising data is not prioritized higher than the service data.

OK

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1160,83 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
+	}
+}
+
+static ssize_t ad_init(struct bt_data *data, const size_t data_size, const atomic_t *adv_opt)
```
Does `data_size` here refer to the number of `bt_data` in the `data` array? Perhaps `size` should be replaced with `count`, `cnt` or `len`.

MariuszSkamra replied:
It refers to the size of `data`, the number of entries that can be added. That about `data` -> `data_array` and `data_size` -> `data_array_size` ?

Thalley replied:
That would work :) 

The term `size` is still a bit unclear, as `size` here, as you've said, refers to the number of entries in `data`, but it could just as easily refer to the actual size (in bytes) of the array. That's why I suggested count or length.

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1147,14 +1160,83 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
 {
 	struct bt_le_adv_param param = {};
-	const struct bt_data *ad;
+	struct bt_data ad[2];
```
`2` seems like a very low number if we want to add flags, service UUIDs and multiple different service data entries.

MariuszSkamra replied:
`cmd_advertise` starts legacy advertisement. Service UUIDs are advertised in extended advertisement only. The RSI is going to be added for legacy advertisement, thus we need 2 entries only (Flags + RSI)

Thalley commented on subsys/bluetooth/shell/csis.c:
```diff
@@ -258,3 +203,36 @@ SHELL_STATIC_SUBCMD_SET_CREATE(csis_cmds,
 SHELL_CMD_ARG_REGISTER(csis, &csis_cmds, "Bluetooth CSIS shell commands",
 		       cmd_csis, 1, 1);
+
+ssize_t csis_ad_data_add(struct bt_data *data, const size_t data_size, const bool discoverable)
```
Looks like this was actually added, but never used :D

MariuszSkamra replied:
It is, in `ad_init`

Thalley replied:
Sorry, must've failed my search last time :D

MariuszSkamra committed:
Bluetooth: audio: Fix undefined symbol compilation error

This fixes compilation error that happen when including audio.h
without CONFIG_BT_AUDIO configuration enabled.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/shell/audio.c:
```diff
@@ -1585,3 +1585,83 @@ static int cmd_audio(const struct shell *sh, size_t argc, char **argv)
+	}
+
+	if (IS_ENABLED(CONFIG_BT_ASCS)) {
+		static uint8_t ad_bap_announcement[8] = {
```
Am I counting wrong, or should this be 
```suggestion
		static uint8_t ad_bap_announcement[7] = {
```
?

MariuszSkamra replied:
Ah, the 8th is metadata length == 0. I will explicitly add it.

Thalley replied:
Ah, that makes sense :)

Thalley reviewed (changes requested):
Minor comment, but otherwise LGTM

Thalley commented on subsys/bluetooth/shell/audio.c:
```diff
@@ -1585,3 +1585,86 @@ static int cmd_audio(const struct shell *sh, size_t argc, char **argv)
+		enum bt_audio_context snk_context, src_context;
+
+		snk_context = bt_audio_capability_get_available_contexts(BT_AUDIO_DIR_SINK);
+		ad_bap_announcement[3] = (((snk_context) >> 0) & 0xFF);
+		ad_bap_announcement[4] = (((snk_context) >> 8) & 0xFF);
```
Consider using something like `sys_put_le16` to insert values in arrays

MariuszSkamra replied:
Ah, yes... Too much copy-paste :D Thanks!
